src/back/msl/writer.rs

```diff
@@ -1265,9 +1213,376 @@ impl<W: Write> Writer<W> {
+                self.put_expression(expr, context, is_scoped)?
+            }
+            index::GuardedIndex::Known(value) => write!(self.out, "{}", value)?,
+        }
+        Ok(())
+    }
+
+    /// Emit an index bounds check condition for `chain`, if required.
+    ///
+    /// `chain` is a subtree of `Access` and `AccessIndex` expressions,
+    /// operating either on a pointer to a value, or on a value directly. If we cannot
+    /// statically determine that all indexing operations in `chain` are within
+    /// bounds, then write a conditional expression to check them dynamically,
+    /// and return true. All accesses in the chain are checked by the generated
+    /// expression.
+    ///
+    /// This assumes that the [`BoundsCheckPolicy`] for `chain` is [`ReadZeroSkipWrite`].
+    ///
+    /// The text written is of the form:
+    ///
+    /// ```ignore
+    /// {level}{prefix}metal::uint(i) < 4 && metal::uint(j) < 10
+    /// ```
+    ///
+    /// where `{level}` and `{prefix}` are the arguments to this function. For [`Store`]
+    /// statements, presumably these arguments start an indented `if` statement; for
+    /// [`Load`] expressions, the caller is probably building up a ternary `?:`
+    /// expression. In either case, what is written is not a complete syntactic structure
+    /// in its own right, and the caller will have to finish it off if we return `true`.
+    ///
+    /// If no expression is written, return false.
+    ///
+    /// [`BoundsCheckPolicy`]: index::BoundsCheckPolicy
+    /// [`ReadZeroSkipWrite`]: index::BoundsCheckPolicy::ReadZeroSkipWrite
+    /// [`Store`]: crate::Statement::Store
+    /// [`Load`]: crate::Expression::Load
+    #[allow(unused_variables)]
+    fn put_bounds_checks(
+        &mut self,
+        mut chain: Handle<crate::Expression>,
+        context: &ExpressionContext,
+        level: back::Level,
+        prefix: &'static str,
+    ) -> Result<bool, Error> {
+        let mut check_written = false;
+
+        // Pull this out into a closure so that we can handle both `Access` and
+        // `AccessIndex` expressions using the same code, using `GuardedIndex` to
+        // represent the index, computed or constant.
+        let mut consider = |base, index| -> BackendResult {
+            if let Some(length) = context.access_needs_check(base, index) {
+                if check_written {
+                    write!(self.out, " && ")?;
+                } else {
+                    write!(self.out, "{}{}", level, prefix)?;
+                    check_written = true;
+                }
+
+                // Check that the index falls within bounds. Do this with a single
+                // comparison, by casting the index to `uint` first, so that negative
+                // indices become large positive values.
+                write!(self.out, "{}::uint(", NAMESPACE)?;
+                self.put_index(index, context, true)?;
+                self.out.write_str(") < ")?;
```

Comment:
let's not mix different styles of writing. Since `write!` is used in other places, let's just use it here as well